cpp/src/lists/contains.cu

```diff
@@ -44,296 +41,241 @@
-    auto result_validity = make_numeric_column(
-      data_type{type_id::BOOL8}, lists.size(), cudf::mask_state::UNALLOCATED, stream, mr);
-    auto mutable_result_positions =
-      mutable_column_device_view::create(result_positions->mutable_view(), stream);
-    auto mutable_result_validity =
-      mutable_column_device_view::create(result_validity->mutable_view(), stream);
-    auto search_key_iter =
-      cudf::detail::make_pair_rep_iterator<ElementType, search_keys_have_nulls>(*d_skeys);
-
-    search_each_list_row<ElementType>(d_lists,
-                                      search_key_iter,
-                                      find_option,
-                                      *mutable_result_positions,
-                                      *mutable_result_validity,
-                                      stream);
-
-    auto [null_mask, num_nulls] = construct_null_mask(lists, result_validity->view(), stream, mr);
-    result_positions->set_null_mask(std::move(null_mask), num_nulls);
-    return result_positions;
+    auto const lists_cdv_ptr = column_device_view::create(lists.parent(), stream);
+    auto const lists_cdv     = cudf::detail::lists_column_device_view{*lists_cdv_ptr};
+    auto const keys_dv_ptr   = get_search_keys_device_view_ptr(search_keys, stream);
+
+    auto out_positions = make_numeric_column(
+      data_type{type_to_id<size_type>()}, lists.size(), cudf::mask_state::UNALLOCATED, stream, mr);
+    auto out_validity   = rmm::device_uvector<bool>(lists.size(), stream);
+    auto const out_iter = thrust::make_zip_iterator(
+      out_positions->mutable_view().template begin<size_type>(), out_validity.begin());
+
+    auto const keys_iter = cudf::detail::make_optional_iterator<Type>(
+      *keys_dv_ptr, nullate::DYNAMIC{search_keys_have_nulls});
+
+    thrust::tabulate(rmm::exec_policy(stream),
+                     out_iter,
+                     out_iter + lists.size(),
+                     search_lists_fn<Type, decltype(keys_iter)>{lists_cdv, keys_iter, find_option});
+
+    if (search_keys_have_nulls || lists.has_nulls()) {
+      auto [null_mask, null_count] = cudf::detail::valid_if(
+        out_validity.begin(), out_validity.end(), thrust::identity{}, stream, mr);
+      out_positions->set_null_mask(std::move(null_mask), null_count);
+    }
+    return out_positions;
   }
 };
 
 /**
- * @brief Converts key-positions vector (from index_of()) to a BOOL8 vector, indicating if
- * the search key was found.
+ * @brief Converts key-positions vector (from `index_of()`) to a BOOL8 vector, indicating if
+ *        the search key(s) were found.
  */
 std::unique_ptr<column> to_contains(std::unique_ptr<column>&& key_positions,
                                     rmm::cuda_stream_view stream,
                                     rmm::mr::device_memory_resource* mr)
 {
   CUDF_EXPECTS(key_positions->type().id() == type_id::INT32,
                "Expected input column of type INT32.");
```

Comment:
We actually interpret this column as `size_type` so the `CUDF_EXPECTS` should mirror that. We should avoid hard-coding the definition that `size_type` is INT32.
```suggestion
  CUDF_EXPECTS(key_positions->type().id() == type_to_id<size_type>,
               "Expected input column of type cudf::size_type.");
```